src/countBy.ts

```diff
@@ -0,0 +1,57 @@
+import { purry } from "./purry";
+
+/**
+ * Categorize and count elements in an array using a defined callback function. The transformation function applies to each element in the array to determine its category and then counts how many elements fall into each category.
+ *
+ * @param data - The array.
+ * @param fn - The transformation function.
+ * @signature
+ *   R.countBy(array, fn)
+ * @example
+ *    R.countBy(
+ *      ["a", "b", "c", "B", "A", "a"],
+ *      (x) => x.toLowerCase()
+ *    ) // { a: 3, b: 2, c: 1 }
+ * @dataFirst
+ * @category Array
+ */
+export function countBy<T, K extends PropertyKey>(
+  data: ReadonlyArray<T>,
+  fn: (value: T, index: number, data: ReadonlyArray<T>) => K,
+): Record<K, number>;
```

Comment:
This should be `ExactRecord` which makes sure that we make the result partial when K is a literal union (a finite set of values).

```suggestion
): ExactRecord<K, number>;
```